src/app/clusters/content-launch-server/content-launch-server.cpp

```diff
@@ -39,32 +39,159 @@
 #include <list>
 
 using namespace chip;
+using namespace chip::app;
+
+// -----------------------------------------------------------------------------
+// Delegate Implementation
+
+using chip::app::Clusters::ContentLauncher::Delegate;
+
+namespace {
+
+Delegate * gDelegateTable[EMBER_AF_CONTENT_LAUNCH_CLUSTER_SERVER_ENDPOINT_COUNT] = { nullptr };
+
+Delegate * GetDelegate(EndpointId endpoint)
+{
+    uint16_t ep = emberAfFindClusterServerEndpointIndex(endpoint, chip::app::Clusters::ContentLauncher::Id);
+    return (ep == 0xFFFF ? NULL : gDelegateTable[ep]);
+}
+
+bool SendStatusIfDelegateNull(EndpointId endpoint)
+{
+    if (GetDelegate(endpoint) == nullptr)
+    {
+        ChipLogError(Zcl, "No ContentLauncher Delegate set for ep:%" PRIu16, endpoint);
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_UNSUP_COMMAND);
+        return true;
+    }
+    return false;
+}
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ContentLauncher {
+
+void SetDelegate(EndpointId endpoint, Delegate * delegate)
+{
+    uint16_t ep = emberAfFindClusterServerEndpointIndex(endpoint, chip::app::Clusters::ContentLauncher::Id);
+    if (ep != 0xFFFF)
+    {
+        gDelegateTable[ep] = delegate;
+    }
+    else
+    {
+    }
+}
+
+} // namespace ContentLauncher
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+// -----------------------------------------------------------------------------
+// Attribute Accessor Implementation
+
+namespace {
+
+class ContentLauncherAttrAccess : public app::AttributeAccessInterface
+{
+public:
+    ContentLauncherAttrAccess() :
+        app::AttributeAccessInterface(Optional<EndpointId>::Missing(), chip::app::Clusters::ContentLauncher::Id)
+    {}
+
+    CHIP_ERROR Read(const app::ConcreteReadAttributePath & aPath, app::AttributeValueEncoder & aEncoder) override;
 
-ContentLaunchResponse contentLauncherClusterLaunchContent(chip::EndpointId endpointId,
-                                                          std::list<ContentLaunchParamater> parameterList, bool autoplay,
-                                                          const chip::CharSpan & data);
+private:
+    CHIP_ERROR ReadAcceptsHeaderAttribute(app::AttributeValueEncoder & aEncoder, Delegate * delegate);
+    CHIP_ERROR ReadSupportedStreamingProtocols(app::AttributeValueEncoder & aEncoder, Delegate * delegate);
+};
 
-ContentLaunchResponse contentLauncherClusterLaunchUrl(const chip::CharSpan & contentUrl, const chip::CharSpan & displayString,
-                                                      ContentLaunchBrandingInformation & brandingInformation);
+ContentLauncherAttrAccess gContentLauncherAttrAccess;
+
+CHIP_ERROR ContentLauncherAttrAccess::Read(const app::ConcreteReadAttributePath & aPath, app::AttributeValueEncoder & aEncoder)
+{
+    EndpointId endpoint = aPath.mEndpointId;
+    Delegate * delegate = GetDelegate(endpoint);
+    if (SendStatusIfDelegateNull(endpoint))
```

Comment:
This is a really odd code pattern (that happens several times).  First we get the delegate from the endpoint.  Then we call a function that gets the delegate from the endpoint and does something with it.  Why not just pass the delegate we already have to that function instead of redoing the "get the delegate" work?

Reply:
Updated. Yeah it does not make sense. It was a copy-pasta from ota-provider-delegate.

Heard once: "Copy/paste is root of all evil"